### PR TITLE
Revision 3 of Online Meetings protoxep

### PR DESCRIPTION
**Pull request by guusdk**

### CODE REVIEW

File: docs/xep/xep-xxxx-http_online_meetings_01-03.xml
Comment (deleolajide) [inline docs/xep/xep-xxxx-http_online_meetings_01-03.xml:1]: Thanks for the review and contribution 👍 

It read better than before and very much compatible with my implementation.

Can we use the  **urn:xmpp:http:online-meetings:xxxxx:0** namespace instead specifying a type attribute. I had removed using the attribute in revision 2. It makes the implementation with Openfire IQ handlers much simpler by having one for each supported provider.

## Example 5
```
<iq from='romeo@montague.tld/garden' 
    id='step_02' 
    to='montague.tld' 
    type='get'>
  <request xmlns='urn:xmpp:http:online-meetings:jitsi:0' id='my-meeting' />
</iq>
````

## Example 6
````
<iq from='montague.tld' 
    id='step_03' 
    to='romeo@montague.tld/garden' 
    type='result'>
  <create xmlns='urn:xmpp:http:online-meetings:jitsi:0'>
    <x xmlns='jabber:x:oob'>
      <url>web+jitsi:https://meet.jit.si/my-meeting-4ba911a9ab321f2d2acd4e67118d5d</url>
      <desc>My Online Meeting</desc>
    </x>
  </create>
  <invite xmlns='urn:xmpp:http:online-meetings:jitsi:0'>
    <x xmlns='jabber:x:oob'>
      <url>web+jitsi:https://meet.jit.si/my-meeting-4ba911a9ab321f2d2acd4e67118d5d</url>
      <desc>My Online Meeting</desc>
    </x>
  </invite>
</iq>
````

## Example 12
````
<message from='romeo@montague.tld/garden'
    to='juliet@capulet.example/balcony'>
  <invite xmlns='urn:xmpp:http:online-meetings:jitsi:0'>
    <x xmlns='jabber:x:oob'>
      <url>https://meet.jit.si/my-meeting-4ba911a9ab321f2d2acd4e67118d5d</url>
      <desc>My Online Meeting</desc>
    </x>
  </invite>
</message>
````

Can we add an additional example for the case where a client entity does not support receiving meeting invitations, but can receive a OOB.

## Example 13
 ````
<message from='romeo@montague.tld/garden'
    to='juliet@capulet.example/balcony'>
    <x xmlns='jabber:x:oob'>
      <url>https://meet.jit.si/my-meeting-4ba911a9ab321f2d2acd4e67118d5d</url>
      <desc>My Online Meeting</desc>
    </x>
</message>
````

Comment (deleolajide) [reply docs/xep/xep-xxxx-http_online_meetings_01-03.xml:1]: As discussed, we also need the requesting entity to specify a room name (or label/description) that is human readable that it creates itself instead. The entity will have a better chance to be using the user's locale, at least.

**Comment by guusdk** [comment]
> Can we use the urn:xmpp:http:online-meetings:xxxxx:0 namespace instead specifying a type attribute. I had removed using the attribute in revision 2. It makes the implementation with Openfire IQ handlers much simpler by having one for each supported provider.

A namespace is used to define that a particular XML element is semantically different from another, that might use the same element name. Here, we're very explicitly re-using the same structure - it doesn't make sense to define them as being 'different', by assigning a different namespace.

I understand that this makes things a bit easier to implement, but that's highly Openfire-specific. We shouldn't compromise the appropriateness of XML definitions for that.

**Comment by guusdk** [comment]
> Can we add an additional example for the case where a client entity does not support receiving meeting invitations, but can receive a OOB.

I do not understand what you're asking.

I've included a directive to add a 'body' element to the message stanza in which the invitation is sent (for clients that do not implement the XEP), but I don't think that this is what you mean.

**Comment by guusdk** [comment]
> As discussed, we also need the requesting entity to specify a room name (or label/description) that is human readable that it creates itself instead. The entity will have a better chance to be using the user's locale, at least.

What I did to include this information, is add an optional 'desc' element to the initiation request. I've chosen to use this name, and an element instead of an attribute, to make it match the 'desc' element of the OOB fragment in which it will be used.

With the names being identical, it should be implicitly obvious that these are the same values. By using the same type (element instead of attribute value), I'm hoping to remove any chance of compatibility issues (eg: using someting silly like a quote-character).

**Comment by guusdk** [comment]
> As discussed later on, maybe we don't need to specify a ID attribute and make that optional.

The ID attribute is now optional. I've added text to explain that this is strictly needed to signal re-usage of an earlier meeting.

File: docs/xep/xep-xxxx-http_online_meetings_01-03.xml
Comment (guusdk) [reply docs/xep/xep-xxxx-http_online_meetings_01-03.xml:1]: I have responded to all these in individual comments.

**Comment by deleolajide** [comment]
> I do not understand what you're asking.

See OOB Example 5 - https://xmpp.org/extensions/xep-0066.html#x-oob

The 'oob' namespace is used to communicate a URI to another user or application. This is done by including an &lt;x/&gt; child element qualified by the 'oob' namespace in either a &lt;message/&gt; and &lt;presence/&gt; stanza; the &lt;x/&gt; child MUST contain a &lt;url/&gt; child specifying the URL of the resource, and MAY contain an optional &lt;desc/&gt; child describing the resource

### Example 5. Sender Communicates a URI

````
<message from='stpeter@jabber.org/work'
         to='MaineBoy@jabber.org/home'>
  <body>Yeah, but do you have a license to Jabber?</body>
  <x xmlns='jabber:x:oob'>
    <url>http://www.jabber.org/images/psa-license.jpg</url>
  </x>
</message>
````
Converse and I think Spark already know how to parse this into an HTML hyperlink. This is a more effective fall-back if the receiving entity cannot handle &lt;invite&gt; element

**Comment by guusdk** [comment]
Is what you're asking already tackled by me having added a directive (and example) to add a 'body' element to the message stanza in which the invitation is sent?

image

If that's not the case, then I'm still not understanding what you're trying to say. :)

**Comment by deleolajide** [comment]
> If that's not the case, then I'm still not understanding what you're trying to say. :)

It is quite close and actually adds the final fall-back for XMPP clients that can't even do HTML. 

The specific fall-back I am asking for is when a receiving entity does not have the capability or support for [urn:xmpp:http:online-meetings:invite:0](http://localhost:7070/xep/xep-xxxx-http_online_meetings_01-03.xml#disco-server), the sending sending entity can in fact use the body as you have shown which will cause the receiving entity to render the URL directly in the conversation thread. Smart clients like Conversations will convert that automatically into a hyperlink.

However, if the OOB fragment is used a direct child node as specified in XEP 0066, then most receiving clients will render the hyperlink today with  the OOB desc as the *content* and the url as the *href* of the &lt;a&gt; HTML tag. 

Example 14 will become
````
<a href="https://meet.jit.si/OpenStandardsMuchGreatness">Meeting room for Open Standards discussion</a>
````

**Comment by deleolajide** [comment]
If the sending entity should confirm if receiving entity supports online meetings. If it does that it sends an &lt;invite&gt; otherwise, it sends an OOB with a &lt;body&gt; fall-back.

**Comment by guusdk** [comment]
> If the sending entity should confirm if receiving entity supports online meetings. If it does then it sends an <invite> otherwise, it sends an OOB with a <body> fall-back.

Why not just always send the fallback, and instruct clients that support the XEP to not show that?

**Comment by deleolajide** [comment]
> Why not just always send the fallback, and instruct clients that support the XEP to not show that?

That is an acceptable approach and I have seen it elsewhere. However, the implementation client side is messy when you are a implementing a plugin.  We could take the OOB out of the invite and make it a sibling. That will allow us to always send both fall-backs.

**Comment by guusdk** [comment]
Isn't that overkill? I'm not against it, but having fallbacks for a fallback ... meh. Maybe just see with the XSF editor / council have to comment on the overall thing?

**Comment by guusdk** [comment]
Isn't the URL exactly how I had it before my last change? I've reverted that commit, and cannot see any diference with what you're proposing now (apart of the nesting of the oob element).

I'm not a fan of moving that oob element to be a sibling, as semantically, that's now no longer related to the invite (and I dislike how the 'invite' now has a different XML Schema than the 'invite' element that's received by the server - that's not very good XMP practice, I think). It is, however, not a hill for me to die on right now (but I suspect that the XSF community will call this out).

**Comment by guusdk** [comment]
(now pushed)

**Comment by deleolajide** [comment]
> apart of the nesting of the oob element

Correct. That is is minor change for backward compatibility. I am happy to duplicate the OOB fragment or even remove it from the invite and add the url and desc elements directly to the invite.

**Comment by guusdk** [comment]
I have now applied those changes. This effectively removes the dependency on XEP-0066 (it's only used as an optional fallback).

**Pull request merged**